qa/tasks/cephfs/test_volumes.py

```diff
@@ -302,5 +356,27 @@ def test_nonexistent_subvolume_group_snapshot_rm(self):
         # remove subvolume
         self._fs_cmd("subvolume", "rm", self.volname, subvolume, group)
 
+        # verify trash dir is clean
+        self._wait_for_trash_empty()
```

Comment:
When I run the tests locally (with vstart_runner on vstart cluster), this test times out,
```
2019-06-21 23:57:14,760.760 INFO:__main__:Running ['./bin/ceph', 'fs', 'flag', 'set', 'enable_multiple', 'true', '--yes-i-really-mean-it']
2019-06-21 23:57:15,912.912 INFO:__main__:Running ['./bin/ceph', 'fs', 'volume', 'ls']
2019-06-21 23:57:16,258.258 INFO:__main__:Running ['./bin/ceph', 'fs', 'subvolumegroup', 'create', u'cephfs', 'group_31']
2019-06-21 23:57:16,665.665 INFO:__main__:Running ['./bin/ceph', 'fs', 'subvolume', 'create', u'cephfs', 'subvolume_9913', '--group_name', 'group_31']
2019-06-21 23:57:17,061.061 INFO:__main__:Running ['./bin/ceph', 'fs', 'subvolumegroup', 'snapshot', 'create', u'cephfs', 'group_31', 'snapshot_9']
2019-06-21 23:57:17,454.454 INFO:__main__:Running ['./bin/ceph', 'fs', 'subvolumegroup', 'snapshot', 'rm', u'cephfs', 'group_31', 'snapshot_9']
2019-06-21 23:57:17,856.856 INFO:__main__:Running ['./bin/ceph', 'fs', 'subvolumegroup', 'snapshot', 'rm', u'cephfs', 'group_31', 'snapshot_9']
2019-06-21 23:57:17.984 7f67000ff700 -1 WARNING: all dangerous and experimental features are enabled.
2019-06-21 23:57:18.053 7f67000ff700 -1 WARNING: all dangerous and experimental features are enabled.
Error ENOENT: Snapshot '/volumes/group_31/.snap/snapshot_9' not found, cannot remove it
2019-06-21 23:57:18,220.220 INFO:__main__:Running ['./bin/ceph', 'fs', 'subvolume', 'rm', u'cephfs', 'subvolume_9913', 'group_31']
2019-06-21 23:57:18,590.590 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:19,622.622 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:20,668.668 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:21,693.693 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:22,719.719 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:23,760.760 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:24,807.807 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:25,855.855 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:26,903.903 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:27,926.926 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:28,971.971 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:30,015.015 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:31,060.060 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:32,107.107 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:33,152.152 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:34,198.198 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:35,222.222 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:36,242.242 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:37,261.261 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:38,280.280 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:39,306.306 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:40,347.347 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:41,363.363 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:42,406.406 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:43,451.451 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:44,497.497 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:45,526.526 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:46,573.573 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:47,590.590 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:48,637.637 INFO:__main__:Running ['getfattr', '--only-values', '-n', 'ceph.dir.entries', '/tmp/tmpuh6UYV/mnt.0/./volumes/_deleting']
2019-06-21 23:57:49,681.681 INFO:__main__:test_nonexistent_subvolume_group_snapshot_rm (tasks.cephfs.test_volumes.TestVolumes) ... ERROR
2019-06-21 23:57:49,682.682 ERROR:__main__:Traceback (most recent call last):
  File "/home/rraja/git/ceph/qa/tasks/cephfs/test_volumes.py", line 360, in test_nonexistent_subvolume_group_snapshot_rm
    self._wait_for_trash_empty()
  File "/home/rraja/git/ceph/qa/tasks/cephfs/test_volumes.py", line 69, in _wait_for_trash_empty
    self.mount_a.wait_for_dir_empty(trashdir)
  File "/home/rraja/git/ceph/qa/tasks/cephfs/mount.py", line 255, in wait_for_dir_empty
    i, dirname, self.client_id))
RuntimeError: Timed out after 30s waiting for ./volumes/_deleting to become empty from 0

2019-06-21 23:57:49,682.682 ERROR:__main__:Error in test 'test_nonexistent_subvolume_group_snapshot_rm (tasks.cephfs.test_volumes.TestVolumes)', going interactive


```
I checked the contents of the filesystem with a manual fuse-mount, the trash folder wasn't empty. It never got purged.

Reply:
I manually populated a subvolume with data, and removed the subvolume. The corresponding trash folder didn't get purged.

Reply:
> I manually populated a subvolume with data, and removed the subvolume. The corresponding trash folder didn't get purged.

what was the uid/gid of the populated data? did you perform a plain `cp`? did you observe any error messages in manager log?

Reply:
Note that the purge thread will fail to remove the directory (and it's entries) from trash if it does not have perms to remove those based on uid/gid. Were the files/dirs created by root?

Reply:
@ajarr ping?

Reply:
@ajarr -- pinging back on this? If this is not an issue then I can post the updated PR for review (minus the fix for assertion).

Reply:
> > I manually populated a subvolume with data, and removed the subvolume. The corresponding trash folder didn't get purged.
> 
> what was the uid/gid of the populated data? did you perform a plain `cp`? did you observe any error messages in manager log?

The uid/gid of the populated data was same as that of the `ceph-mgr` process.

Reply:
I added more logging and the d_name.decode() fix, 
```
diff --git a/src/pybind/mgr/volumes/fs/purge_queue.py b/src/pybind/mgr/volumes/fs/purge_queue.py
index e1e9dc1dae..81aa4bb436 100644
--- a/src/pybind/mgr/volumes/fs/purge_queue.py
+++ b/src/pybind/mgr/volumes/fs/purge_queue.py
@@ -80,6 +80,8 @@ class PurgeQueueBase(object):
 
         thread_id = threading.currentThread()
         self.jobs[volname].append((purge_dir, thread_id))
+        log.debug("In register_job: job for volume {0} is {1}".format(volname, self.jobs[volname]))
+
 
     def unregister_job(self, volname, purge_dir):
         log.debug("unregistering purge job: {0}.{1}".format(volname, purge_dir))
@@ -109,6 +111,7 @@ class PurgeQueueBase(object):
 
     def purge_trash_entry_for_volume(self, volname, purge_dir):
         thread_id = threading.currentThread()
+        log.debug("In purge_trash_entry_for_volume: vol name '(0}' and purge dir '{1}'".format(volname, purge_dir))
         ret = self.vc.purge_subvolume_trash_entry(
             None, vol_name=volname, purge_dir=purge_dir, should_cancel=lambda: thread_id.should_cancel())
         return ret[0]
@@ -140,6 +143,7 @@ class ThreadPoolPurgeQueueMixin(PurgeQueueBase):
             # for next volume.
             self.q.rotate(1)
             ret, purge_dir = self.get_trash_entry_for_volume(volname)
+            log.debug("In pick_purge_dir_from_volume: purge dir is '{0}'".format(purge_dir))
             if purge_dir:
                 to_purge = volname, purge_dir
                 break
diff --git a/src/pybind/mgr/volumes/fs/subvolume.py b/src/pybind/mgr/volumes/fs/subvolume.py
index 9b86a510ac..cd349f5046 100644
--- a/src/pybind/mgr/volumes/fs/subvolume.py
+++ b/src/pybind/mgr/volumes/fs/subvolume.py
@@ -71,8 +71,8 @@ class SubVolume(object):
         d = self.fs.readdir(dir_handle)
         d_name = None
         while d:
-            if not d.d_name in exclude and d.is_dir():
-                d_name = d.d_name
+            if not d.d_name.decode('utf-8') in exclude and d.is_dir():
+                d_name = d.d_name.decode('utf-8')
                 break
             d = self.fs.readdir(dir_handle)
         self.fs.closedir(dir_handle)
diff --git a/src/pybind/mgr/volumes/fs/volume.py b/src/pybind/mgr/volumes/fs/volume.py
index e69d35ad4f..64f0690026 100644
--- a/src/pybind/mgr/volumes/fs/volume.py
+++ b/src/pybind/mgr/volumes/fs/volume.py
@@ -550,6 +550,7 @@ class VolumeClient(object):
             with SubVolume(self.mgr, fs_handle) as sv:
                 spec = SubvolumeSpec("", "")
                 path = sv.get_trash_entry(spec, exclude)
+                log.debug("In get_subvolume_trash_entry: trash path is '{0}'".format(path))
                 ret = 0, path, ""
         except VolumeException as ve:
             ret = self.volume_exception_to_retval(ve)
```

Even though a thread picks up the trashdir and registers it, the thread doesn't seem to purge it, i.e., execute `purge_trash_entry_for_volume`. See the mgr log snippet, after the `fs subvolume rm` of a subvolume is called,
https://pastebin.com/PyNHqrB6

Reply:
> Even though a thread picks up the trashdir and registers it, the thread doesn't seem to purge it, i.e., execute `purge_trash_entry_for_volume`. See the mgr log snippet, after the `fs subvolume rm` of a subvolume is called, https://pastebin.com/0MWqCJG7 . I didn't see the debug log message of being in purge_trash_entry_for_volume in the mgr log.

that may be due to buffered logging (I've seen that before in an mgr module). if an exception is raised in `purge_trash_entry_for_volume()`, the string is never logged.